File: 0049_Group_Anagrams/0049_Group_Anagrams.swift
class Solution {
    func groupAnagrams(_ strs: [String]) -> [[String]] {
        var groupedAnagrams: [String: [String]] = [:]

        // Read str from strs
        for elem: String in strs {
            // Convert to sorted string
            let elemStr:String = String(elem.sorted())
            
            // Anagram group exists, append to existing group
            if groupedAnagrams.keys.contains(elemStr) {
                if var anagrams: [String] = groupedAnagrams[elemStr] {
                    anagrams.append(elem)
                    groupedAnagrams[elemStr] = anagrams
                }
            }
            // Anagram Group Does not exits, start a new group
            else {
                groupedAnagrams[elemStr] = [elem]
            }
        }

        return Array(groupedAnagrams.values)
    }
}

File: 0049_Group_Anagrams/0049_Group_Anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
    const groupedAnagrams= new Map<string, string[]>();

    // Read all elems from strs
    for (let elem of strs) {
        // Sort elem by chracter
        const sortedElem = [...elem].sort().join('');

        // Anagram Group Exists: append to new group
        if (groupedAnagrams.has(sortedElem)) {
            const anagrams: string[] = groupedAnagrams.get(sortedElem) ?? [];
            groupedAnagrams.set(sortedElem, [...anagrams, elem]);
        }
        // Anagram Group Does Not Exist
        else {
            groupedAnagrams.set(sortedElem, [elem]);
        }
    }

    return Array.from(groupedAnagrams, ([key, value]) => value);
};